app_dart/lib/src/request_handlers/get_green_commits.dart

```diff
@@ -0,0 +1,57 @@
+// Copyright 2021 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:cocoon_service/src/model/appengine/stage.dart';
+import 'package:cocoon_service/src/model/appengine/task.dart';
+import 'package:github/github.dart';
+import 'package:meta/meta.dart';
+
+import '../request_handling/body.dart';
+import '../request_handling/request_handler.dart';
+import '../service/build_status_provider.dart';
+import '../service/config.dart';
+import '../service/datastore.dart';
+
+@immutable
+class GetGreenCommits extends RequestHandler<Body> {
+  const GetGreenCommits(
+    Config config, {
+    @visibleForTesting this.datastoreProvider = DatastoreService.defaultProvider,
+    @visibleForTesting BuildStatusServiceProvider? buildStatusProvider,
+  })  : buildStatusProvider = buildStatusProvider ?? BuildStatusService.defaultProvider,
+        super(config: config);
+
+  final DatastoreServiceProvider datastoreProvider;
+  final BuildStatusServiceProvider buildStatusProvider;
+
+  static const String kRepoParam = 'repo';
+
+  @override
+  Future<Body> get() async {
+    final String repoName = request!.uri.queryParameters[kRepoParam] ?? Config.flutterSlug.name;
+    final RepositorySlug slug = RepositorySlug('flutter', repoName);
+    final String branch = Config.defaultBranch(slug);
+    final DatastoreService datastore = datastoreProvider(config.db);
+    final BuildStatusService buildStatusService = buildStatusProvider(datastore);
+    final int commitNumber = config.commitNumber;
+    final int lastCommitTimestamp = DateTime.now().millisecondsSinceEpoch;
+
+    final List<String?> greenCommits = await buildStatusService
+        .retrieveCommitStatus(
+          limit: commitNumber,
+          timestamp: lastCommitTimestamp,
+          branch: branch,
+          slug: slug,
+        )
+        .where((CommitStatus status) => status.stages.every((Stage s) => s.taskStatus == Task.statusSucceeded))
+        .map<String?>((CommitStatus status) => status.commit.sha)
+        .toList();
+
+    return Body.forJson(<String, List<String?>>{
+      'greenCommits': greenCommits,
```

Comment:
Why return a Map? You can just return a List<String> which is easier to use

Reply:
umm Maybe my understanding is not correct, but I think the get() methods in request handlers have to be a valid override of the `Future<Body> get()` function, which requires a `Future<Body>` as a return type. And to create a Body type, the three valid ways are through `Body.forString`, `Body.forStream` and `Body.forJson`, which requires input type as String, Stream<Uint8List?> and dynamic(json) respectively. I am not sure how to cast List<String> into a String or Stream<Uint8List?> type, and so I ended up making it a valid json type (a map, sort of), and then feed it into Body.forJson to create a Body type. umm Would there be a valid creation of Body type if we use List<String> type?

Reply:
List<String> is valid json. You can use Body.fromJson(listStringVar)

Reply:
Thanks for the hint! I have now casted Stream.single to List<dynamic> and then to List<String> to make it work for List type.